---
output:
  html_document:
    code_folding: hide
    theme: yeti
    toc: yes
    toc_float: true
    toc_depth: 6
    number_sections: true
    fig_height: 6 
title: "Marneros - Knockout RNAseq Analysis"
bibliography: references.bib
author: "John N Hutchinson"
---

[R]: https://www.r-project.org
[RStudio]: https://www.rstudio.com


[bcbio-rnaseq]: https://github.com/roryk/bcbio.rnaseq
[bcbio]: https://github.com/chapmanb/bcbio-nextgen
[DESeq2]: https://bioconductor.org/packages/release/bioc/html/DESeq2.html
[salmon]: https://combine-lab.github.io/salmon/
[tximport]: https://bioconductor.org/packages/release/bioc/html/tximport.html
[Bioconductor]: https://bioconductor.org
[Ensembl]: http://useast.ensembl.org/Mus_musculus/Info/Index




```{r setup1, echo=FALSE}
# Setup report details
clientname="Alex Marneros"
clientemail="amarneros@mgh.harvard.edu"
lablocation="MGH" 
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```

> RNA-Seq analysis for `r clientname` (`r clientemail`) at `r lablocation`. Contact `r analystname` (`r analystemail`) for additional details. 
> This document was generated on `r date()` using [R][] and [RStudio][]
 
```{r setup2, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

Fastq files were processed in [bcbio][] and pseudo-count data generated by [salmon][]. Various RNAseq metrics were summarized with [bcbio-rnaseq][]. 

> The project has 8 samples, 4 controls and 4 cell specific Kcnd1 knockouts.

Here I do some basic QC to see if the data looks good and if whether any samples look like they have issues.

```{r load-libraries, echo=FALSE}
check_install = function(packages) {
   not_installed = setdiff(packages, rownames(installed.packages()))
   if(length(not_installed) > 0) {
      write(paste("The libraries", not_installed, "are not installed, aborting.",
                   sep=" "), stderr())
      stop()
   }
}
packages = c("ggplot2", "reshape", "gplots", "edgeR", "CHBUtils", "pheatmap","DESeq2", "tximport", "DT", "DEGreport", "dplyr", "janitor", "rio", "rdrop2")
check_install(packages)
installed = lapply(packages, library, character.only = TRUE)
```

```{r qc-setup, results='hide'}
project_summary = "/Volumes/orch/PIs/alex_marneros/am_rnaseq/bcbio/final/2017-04-14_bcbio/project-summary.csv"
counts_file = "/Volumes/orch/PIs/alex_marneros/am_rnaseq/bcbio/final/2017-04-14_bcbio/combined.counts"
tx2genes_file = "/Volumes/orch/PIs/alex_marneros/am_rnaseq/bcbio/final/2017-04-14_bcbio/tx2gene.csv"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442","#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
summarydata = summarydata[,colSums(is.na(summarydata)) < nrow(summarydata)]
# handle newer bcbio-nextgen runs that use description as the key
if("description" %in% colnames(summarydata)) {
  rownames(summarydata) = summarydata$description
  summarydata$Name = rownames(summarydata)
  summarydata$description = NULL
} else {
  rownames(summarydata) = summarydata$Name
  # summarydata$Name = NULL
}
summarydata = summarydata[order(rownames(summarydata)),]
colnames(summarydata) = gsub(".", "_", colnames(summarydata), fixed=TRUE)

summarydata <- clean_names(summarydata)
summarydata$name <- summarydata$descrption
row.names(summarydata) <- summarydata$name


if (file.exists(tx2genes_file)) {
  sf_files = list.files(file.path(dirname(tx2genes_file), "../../"), pattern="quant.sf", recursive=TRUE, full.names=TRUE)
  if(length(sf_files) > 0) {
    sf_files = sf_files[order(sf_files)]
    names(sf_files) = rownames(summarydata)
    tx2gene = read.table(tx2genes_file, sep=",", row.names=NULL, header=FALSE)
    txi.salmon = tximport(sf_files, type="salmon", tx2gene=tx2gene,
             reader=readr::read_tsv, countsFromAbundance="lengthScaledTPM")
    counts = round(data.frame(txi.salmon$counts, check.names=FALSE))
} else {
  counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
}}
counts = counts[, order(colnames(counts)), drop=FALSE]
names(counts) <- row.names(summarydata)

summarydata = summarydata[order(rownames(summarydata)),]
counts = counts[, order(colnames(counts)), drop=FALSE]

# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("rrna_rate", "duplication_rate_of_mapped", "mapped_reads", "mapped_paired_reads", "quality_format", "intergenic_rate", "sequence_length", "exonic_rate", "total_reads", "percentgc", "rrna", "duplicates", "name", "sequences_flagged_as_poor_quality", "average_insert_size", "intronic_rate", "x5_3_bias", "fragment_length_mean","x_gc", "mapped_reads_pct")
summarydata[,"mapped_reads_pct"] <- (summarydata$mapped_reads/summarydata$total_reads)*100

metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
metadata = metadata[, colSums(is.na(metadata)) < nrow(metadata), drop=FALSE]
metadata <- metadata[, c("replicate", "genotype")] 
```


```{r functions, echo=FALSE}
#pca-plotter
pca_plot = function(comps, nc1, nc2, colorby, shapeby) {
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  if(!(c1str %in% colnames(comps) && c2str %in% colnames(comps))) {
    warning("Higher order components not found, skipping plotting.")
    return(NA)
  }
  ggplot(comps, aes_string(c1str, c2str, color=colorby, shape=shapeby)) +
    geom_point(size=5) + theme_bw() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
  }

#sanitize-datatable
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}

#heatmap-function
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, clustering_method = "ward.D2",
        clustering_distance_cols = "correlation", ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)

annotate_df2 <- function (df, df_ensemblid_header, biomart_ensembl_dataset, biomart_ensemblid_filter, biomart_genesymbol_attribute, biomart_host) {
  require(biomaRt)
  ensembl = useMart("ENSEMBL_MART_ENSEMBL", dataset = biomart_ensembl_dataset, host = biomart_host)
  annot.df = getBM(attributes = c(biomart_ensemblid_filter, biomart_genesymbol_attribute, "description"), filters = c(biomart_ensemblid_filter), values = as.character(df[, df_ensemblid_header]), mart = ensembl)
  m = merge(df, annot.df, by.x = df_ensemblid_header, by.y = biomart_ensemblid_filter,all.x = T)
  return(m)
}

subset_tximport = function(txi, rows, columns) {
  txi$counts = txi$counts[rows, columns]
  txi$abundance = txi$abundance[rows, columns]
  txi$length = txi$length[rows, columns]
  return(txi)
}
DESeqDataSetFromTximport <- function(txi, colData, design, ...) {
  counts <- round(txi$counts)
  mode(counts) <- "integer"
  dds <- DESeqDataSetFromMatrix(countData=counts, colData=colData, design=design, ...)
  stopifnot(txi$countsFromAbundance %in% c("no","scaledTPM","lengthScaledTPM"))
  if (txi$countsFromAbundance %in% c("scaledTPM","lengthScaledTPM")) {
    message("using length scaled TPM counts from tximport")
  } else {
    message("using counts and average transcript lengths from tximport")
    lengths <- txi$length
    dimnames(lengths) <- dimnames(dds)
    assays(dds)[["avgTxLength"]] <- lengths
  }
  return(dds)
}

```

```{r set-random-seed}
# set seed for reproducibility
set.seed(1454944673)
```

## Sample metadata
```{r summary-table, echo=FALSE, cache=FALSE}
sanitize_datatable(summarydata, style='bootstrap')
```


# Quality control metrics

```{r qualimap-run}
qualimap_run = "exonic_rate" %in% colnames(summarydata)
do_quality = "total_reads" %in% colnames(summarydata)
```

## Mapped reads

The number of mapped reads is a bit low (we'd prefer to see >20 million reads mapped) but still perfectly within an acceptable range.

```{r mapped-plot, eval=do_quality}
ggplot(summarydata, aes(x=name, y=mapped_reads, fill=genotype)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```


## Genomic mapping rate

The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts. We don't see that here. The mapping rate is acceptable (>90%) for all samples:

```{r mapping-rate-plot, eval=do_quality}
mapping_string = "mapped_reads_pct"
ggplot(summarydata, aes_string(x="name", y=mapping_string, fill="genotype")) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```


## Number of genes detected

The number of genes detected for each sample is good with over 20,000 genes detected for all samples. This is what we expect for the mouse genome:

```{r genes-detected-plot}
dd = data.frame(Name=colnames(counts), genes_detected = colSums(counts > 0), genotype=summarydata$genotype)
ggplot(dd, aes(x=Name, y=genes_detected, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation

Gene detection saturation was nearly reached for the majority of the samples, which means that most of the samples were able to detect the majority of expressed genes. This is good to see. There is no linear trend here of genes detected increasing with the number of reads mapped ine the samples, which may actually indicate sample overloading.

```{r saturation-plot, eval=do_quality}
col_mapped =  "mapped_reads"
dd = data.frame(mapped=summarydata[,col_mapped], genes_detected = colSums(counts > 0), genotype=summarydata$genotype)
ggplot(dd, aes(x=mapped, y=genes_detected, color=genotype)) +
    geom_point(size=5) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped")
```


## Exonic mapping rate

The exonic mapping rates are good, with over 60% of total reads mapping to exons.

```{r exonic-mapping-plot, eval="exonic_rate" %in% colnames(summarydata)}
ggplot(summarydata, aes(x=name, y=exonic_rate, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```


## rRNA mapping rate

All samples are more than within an acceptable range (<10%).

```{r check-rRNA-eval}
eval_rRNA = "rrna_rate" %in% colnames(summarydata) & !sum(is.na(summarydata$rrna_rate)) == nrow(summarydata)
```
```{r rRNA-rate-plot, eval=eval_rRNA & do_quality}
ggplot(summarydata, aes(x=name, y=rrna_rate, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

These values are fine. Anything below 5% is reasonable, the KO_2 sample is relatively higher but in absolute terms, sill quite low.

## 5'->3' bias

There is some bias, but it is fairly consistent among samples.

```{r 53bias, eval="x5_3_bias" %in% colnames(summarydata) & do_quality}
ggplot(summarydata, aes(x=name, y=x5_3_bias, fill=genotype)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("5'->3' bias") + xlab("")
```

## Boxplot of log10 counts per gene

Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different. The count spreads look good for all samples:

```{r boxplot-raw}

melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of normalized counts per gene

The DESeq2  [DESeq2][] [@Love2014] base normalization method adjusts the counts by library size. By normalizing in this way for total RNA expression by sample, we expect the spread of the log<sub>10</sub> TMM-normalized counts per gene to be similar for every sample. The samples look good, with similar log<sub>10</sub> DESeq2 normalized counts per gene distributions:

```{r boxplot-normalized}
counts <- counts[rowSums(counts>0)>1,]
txi.salmon = subset_tximport(txi.salmon, rownames(counts), colnames(counts))
dds = DESeqDataSetFromTximport(txi.salmon, colData=summarydata, design=~1)

geoMeans = apply(counts, 1, function(row) if (all(row == 0)) 0 else
                 exp(mean(log(row[row != 0]))))
dds = estimateSizeFactors(dds, geoMeans=geoMeans)
dds = DESeq(dds, betaPrior=TRUE)

normalized_counts <- as.data.frame(counts(dds, normalized=TRUE))
normalized_counts$id <- as.character(rownames(normalized_counts))


melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of normalized counts

Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different. The samples look good, with similar log<sub>10</sub> DESeq2-normalized counts per gene densities:

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

# Knockout Gene Expression

We can use the DESeq2 normalized counts to look at the expression of the knockout gene in the samples.

```{r dds_setup, results="hide"}
# annotate to get the gene symbols
normcounts_annot <- annotate_df2(df=normalized_counts, df_ensemblid_header = "id", biomart_ensembl_dataset= 'mmusculus_gene_ensembl',biomart_ensemblid_filter = "ensembl_gene_id", biomart_genesymbol_attribute = 'mgi_symbol', biomart_host="useast.ensembl.org" )

# melt data to prep for dplyr tomfoolery
normcounts_annot <- reshape2::melt(normcounts_annot, id.vars=c("mgi_symbol", "id", "description")) 
normcounts_annot <- normcounts_annot %>% tbl_df()

# merge with metagdata to get genotypes
summarydata_dt <- tbl_df(summarydata)
normcounts_annot <- inner_join(normcounts_annot, summarydata_dt, by = c("variable" = "name"))

# grab just the gene expression we need
plotme <- filter(normcounts_annot, mgi_symbol=="Kctd1") %>% dplyr::select(mgi_symbol, value, genotype, variable)

# plot
ggplot(plotme, aes(x=variable,y=value, fill=genotype ))+geom_bar(stat="identity")+theme_bw()
ggplot(plotme, aes(x=genotype, y=value, color=genotype))+geom_boxplot()+geom_point(data=plotme, aes(y=value, x=genotype, fill=genotype), size=3) +theme_bw()
```

We are definitely seeing a reduction in Kcnd1 expression in the knockout samples.

# Sample Similarity Analyses (Unsupervised clustering)

## Transform counts to log<sub>2</sub>

Before performing similarity analysis, we transform counts to log<sub>2</sub>, which acts to minimize large differences in sequencing depth and helps normalize all samples to a similar dynamic range.

For RNA-seq count data, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log<sub>2</sub> transformation will amplify this noise, and show the low count genes will dominate the PCA plots. For high counts, this gives similar results as log<sub>2</sub>, but for low counts, values are shrunken towards the genes' average across samples.

This is performed with `rlog` function in the [DESeq2][] package [@love2014deseq2], which we will later use for differential gene expression analysis.

```{r ddsvst}
rld <- rlog(dds)
rlogMat <- assay(rld)
```


## Correlation heatmap of TMM-normalized counts {.tabset}

We aren't seeing particularly good correlations within sample classes, which suggests some variance in the data.

### Correlation (Pearson)
```{r pearson-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(rlogMat, method="pearson"))
```

### Correlation (Spearman)
```{r spearman-heatmap-normalized, eval=ncol(counts) > 1}
heatmap_fn(cor(rlogMat, method="spearman"))
```

## PCA plots {.tabset}

The PCA plots
```{r pca, eval=ncol(counts) > 1}
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design=~name)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(summarydata, by=c("name"="name"))
colorby = "genotype"
comps$replicate <- as.character(comps$replicate)
shapeby = "replicate"
```

### PC1 vs. PC2
```{r PC1-PC2}
pca_plot(comps, 1, 2, colorby, shapeby)+scale_shape_manual(values=c(15:18)) 
```

### PC3 vs. PC4
```{r PC3-PC4}
pca_plot(comps, 3, 4, colorby, shapeby)+scale_shape_manual(values=c(15:18)) 
```

### PC5 vs. PC6
```{r PC5-PC6}
pca_plot(comps, 5, 6, colorby, shapeby)+scale_shape_manual(values=c(15:18)) 
```

### Variance explained by component
```{r component-loadings}
ggplot(data.frame(component=reorder(names(pc$percentVar), -pc$percentVar),
                  percent_var=pc$percentVar),
       aes(component, percent_var)) + geom_bar(stat='identity') +
  ylab("percent of total variation") + xlab("") + theme_bw()
```
Â 



---


# SessionInfo

```{r sessioninfo}
sessionInfo()
```


# References