```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, fig.height=8, fig.width=8, figalign="center", echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120), dev='CairoSVG')
options(width=200)

clientname="Rohan Manohar"
clientemail="rohanraoul@gmail.com"
labPI="Wagers"
lablocation="Joslin"
analystname="John Hutchinson"
analystemail="jhutchin@hsph.harvard.edu"
```
---

RNAseq (DESeq) analysis for `r clientname` (`r clientemail`), `r labPI` group at `r lablocation`.  Contact `r analystname` (`r analystemail`) for additional details.

The most recent update of this html document occurred: `r date()`

---

## Methods summary

All reads were trimmed to remove adapter sequences and low quality calls.  Trimmed reads were aligned with [Tophat2][1] and gene specific read counts for [Ensembl gene annotations][2] determined with [htseq-count][3]. Read counts were then normalized and differential gene expression calculated with the [edgeR][4] Bioconductor package.

The sections below provide code to reproduce the post-counting results and plots. 

---

## SETUP

### Libraries

Bioconductor and R libraries used to process and visualize the data

```{r libraries}
library(DESeq)
library(plyr)
library(reshape)
library(ggplot2)
library(biomaRt)
library(scales)
library(googleVis)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")
library(grid)
library(VennDiagram)
library(OrderedList)
```

### Variables

Working directories, cutoffs and other variables necessary to the analysis.

```{r variables }
if (file.exists("/n/hsphS10/hsphfs1/chb/projects/aw_rnaseq_dev/")) {
  baseDir <- "/n/hsphS10/hsphfs1/chb/projects/aw_rnaseq_dev/"
} else {
  baseDir <- "/Volumes/ody_new/projects/aw_rnaseq_dev/"
}
dataDir <- file.path(baseDir, "htseq-count/")
resultsDir <- file.path(baseDir, "results/deseq")
metaDir <- file.path(baseDir, "meta")
count.file <- file.path(dataDir, "combined.counts")

pvalcutoff=0.001
numsig=10
rowcount.cutoff=10
logfoldcutoff=1

gene_symbol = 'mgi_symbol'
ensembl_gene = 'mmusculus_gene_ensembl'
filter_type = 'ensembl_gene_id'
```

For this analysis, the cutoff for significant differential expression was an FDR (False Discovery Rate) of `r pvalcutoff` and a log2(foldchange) of expression of at least `r logfoldcutoff`.  

### Functions

```{r functions}

source("http://dl.dropboxusercontent.com/u/4253254/Resources/functions.r")

annotate_df = function(d) {
  require(biomaRt)
	ensembl = useMart('ensembl', dataset = ensembl_gene)
	a = getBM(attributes=c(filter_type, gene_symbol, "description"),
		filters=c(filter_type), values=d[, 'id'],
		mart=ensembl)
	m = merge(d, a, by.x='id', by.y=filter_type)
	return(m)
}

plotDispEsts = function(cds) {
  estimates = data.frame(means = rowMeans(counts(cds, normalized=TRUE)),
		variance = fitInfo(cds)$perGeneDispEsts)
	xg = 10^seq(-0.5, 5, length.out=300)
	yg = fitInfo(cds)$dispFun(xg)
	fitline = data.frame(xg=xg, yg=yg)
	p = ggplot(estimates, aes(means, variance)) + geom_point(size=1, alpha=0.4) +
		scale_x_log10() + scale_y_log10() +
		geom_line(data=fitline, aes(xg, yg), color="red") +
		labs(title="dispersion estimation while pooling all samples") +
		xlab("mean number of mapped reads per gene") +
		ylab("estimated dispersion")
	p
}

plotDE <- function(res, cutoffpval, samples) {
  title = paste("M-A plot of", samples, sep = " - ")
  res$colors <- ifelse(res$padj < cutoffpval, "sig", "nonsig")
  plot <- ggplot(data = res, aes(x = log(baseMean), y = log2(foldChange), colour = colors)) + 
    geom_point(size = 3) + 
    scale_colour_manual(name = "BFH adjusted pvalue", values = c("#00000033", "#FF0000FF"), labels = c(paste("q>", cutoffpval, sep = ""), paste("q<", cutoffpval, sep = ""))) + 
    labs(title = title)
  plot
}

qq = function(pvaldf,  cutoffpval, samples) {
  title=paste("Quantile-quantile plot of p-values", samples, sep=" - ")
  pvaldf <- pvaldf[order(pvaldf$pval, decreasing=F),]
  pvals <- as.vector(unlist(pvaldf$pval))
  padjs <- as.numeric(as.vector(unlist(pvaldf$padj)))
  colors <- as.vector(ifelse(padjs<cutoffpval, "sig", "nonsig"))
  o = -log10(pvals)
  e = -log10( 1:length(o)/length(o) )
  plot=qplot(e,o, color=colors, xlim=c(0,max(e[!is.na(e)])), ylim=c(0,max(o[!is.na(o)]))) + stat_abline(intercept=0,slope=1, col="darkgrey")
  plot=plot+labs(title=title)
  plot=plot+scale_x_continuous(name=expression(Expected~~-log[10](italic(p))))
  plot=plot+scale_y_continuous(name=expression(Observed~~-log[10](italic(p))))
  plot=plot + scale_colour_manual(name="BFH adjusted pvalue", values=c("black", "red"), labels=c(paste("q>", cutoffpval, sep=""),paste("q<", cutoffpval,sep=""))) 
  plot
}


pcaPlot <- function(countdataset, title, palette)  {
  fit <- prcomp(t(countdataset))
  colors <- palette[factor(pData(cds)$condition)]
  legend_values=unique(cbind(colors, as.character(pData(cds)$condition)))
  ##all samples
  plot(fit$x, bg=colors, col="black", cex=2,pch=21, main=title, oma=c(8,5,5,14))
  legend("topright", cex=0.7, col="black", pt.bg=legend_values[,1], pt.cex=1.25, legend=legend_values[,2],  pch=21, bty="n", x.intersp=1)
  }

PCAplot.cds <- function(countdataset=NULL, metadata=NULL, categories=NULL, title=NULL, colorpalette=NULL, alpha=1, numcomponents=6, normalize.counts=TRUE){
  alpha <- sprintf("%x", ceiling(alpha*255))
  colorpalette <- paste(colorpalette, alpha, sep="")
  if (normalize.counts){
    df <- counts(countdataset, normalized=TRUE)
    } else {
      df <- counts(countdataset, normalized=FALSE)
      }
  myPca.core <- prcomp(t(df))
  tmpPCAData.core <- as.data.frame(myPca.core$x[,1:numcomponents])
  colors <- colorpalette[factor(countdataset$condition)]
  legend_values=unique(cbind(colors, as.character(countdataset$condition)))
  pairs(tmpPCAData.core, bg=colors, col="#606060", cex=2, pch=21, main=title, oma=c(8,5,5,14))
  legend("right", cex=0.7, col="#606060", pt.bg=legend_values[,1], pt.cex=1.5, legend=legend_values[,2],  pch=21, bty="n", x.intersp=1)
  }
```

--- 

## DATA IMPORT

The data consists of HTseq counts generated for individual genes (using the Ensembl GTF annotation) from the Tophat aligments. The metadata (or covariates) that describes the experiment consists of the  replicated experiemental classes of "FAPLung", "FAPMuscle" and "MEF"; with 4 (FAPMuscle and MEF) or 5 (FAPLung) replicates per time point. 

These data and metadata were and loaded into a DESeq *CountDataSet*.

```{r dataload_and_reshape}
combined.counts <- read.table(file.path(dataDir, "combined.counts"), header=T, row.names=1)
combined.counts <- combined.counts[, order(names(combined.counts), decreasing=F)]
# use samplenames get covars from the filenames of the counted samples
# identical values in this vector will be marked as replicates
covars <- factor(sub("[0-9]", "", names(combined.counts)))
## load up new count dataset
cds <- newCountDataSet(combined.counts, covars)
```

---
 
## NORMALIZATION

Here is a sample of `r numsig` of the non-normalized counts of reads mapping to
genes in the annotation. These are just chosen at random for display purposes:

```{r rawcounts, results='asis'}
annotated_counts = head(as.data.frame(counts(cds)), numsig)
annotated_counts <- rownames2col(annotated_counts, "id")
annotated_counts = annotate_df(annotated_counts)
annotated_counts.gvis <- gvisTable(as.data.frame(apply(annotated_counts, 2, as.character)), options=list(width=1280))
```

```{r printmetest, results='asis'}
print(annotated_counts.gvis,  "chart")
```

To normalize for different depth of sequencing, size factors are calculated for each sample. 

Call $k_{ij}$ the count for gene $i$ in sample $j$. 

For all samples $j$, assume most of the counts for genes will have a constant ratio comparing any two samples in $j$, reflecting the assumption that most genes are not differentially expressed between samples. 

A reference set of counts is created by taking $f_i$, geometric mean of the counts for each gene $i$, and the median of of $\frac{k_i}{f_i}$ for sample $j$ is called the size factor for sample $j$. 

The size factors can be thought of as an attempt to normalize for the total number of reads mapping to transcripts, depending on the assumption that most of the transcripts are not vastly different between samples.

Here are the size factors for each sample in this experiment:

```{r sizefactors, tidy=TRUE}
cds = estimateSizeFactors(cds)
sizeFactors(cds)
```

These are the normalized counts, which are the raw counts divided by the size factor, for the same `r numsig` genes:

```{r normalized_counts, results='asis'}
annotated_normalized = head(as.data.frame(counts(cds, normalized=TRUE)), numsig)
annotated_normalized  <- rownames2col(annotated_normalized, 'id')
annotated_normalized = annotate_df(annotated_normalized)
annotated_normalized.gvis <- gvisTable(as.data.frame(apply(annotated_normalized, 2, as.character)), options=list(width=1280))
print(annotated_normalized.gvis, "chart")
```

We can see the assumption holds true by plotting the ratio of counts for any two randomly drawn samples and noting the distribution has a large central peak.

In blue are the raw counts and in red are the normalized counts. The effect of normalization is to shift the mean ratio towards 1.

Here are a couple examples:

```{r ratio_hist, out.width='50%', results='hide', fig.show='hold'}
raw.counts = counts(cds, normalized=FALSE)
cols <- sample(ncol(raw.counts),4, replace=F)
norm.counts = counts(cds, normalized=TRUE)

for(x in c(1,3)){
  rawdata = data.frame(ratio=raw.counts[,cols[x]] / raw.counts[,cols[x+1]])
  rawdata$set <- "raw"
  
  normdata = data.frame(ratio=norm.counts[,x] / norm.counts[,x+1])
  normdata$set <- "normalized"
  raw.norm.data <- rbind(rawdata, normdata)
  
  n = ggplot(raw.norm.data, aes(x=ratio, fill=set)) + geom_density(alpha=0.25) +
    scale_x_log10(breaks=c(0.01, 0.1, 1, 10, 100), labels=math_format(format=log10)) +
    labs(title="Normalized counts")
  print(n)
}
rm(raw.counts, norm.counts)
```

## Variance estimation

With replicates you can get a more accurate repesentation of the biological variability or dispersion. Dispersion describes the fluctuation of the observed counts for a gene around the mean counts for that gene and can be the square of the coefficient of biological variation. For example, if a gene's expression level fluctuates by 20% the calculated dispersion is $0.2^2 = 0.04$.

Below is a plot of the estimated dispersion for genes with given counts; the dispersion increases as the mean count decreases meaning it is more difficult to accurately measure the expression level of
low-count genes.

```{r estimate_sizefactors, results='hide', fig.cap="Empirical and fitted dispersion values plotted against mean expression strength", out.width='50%'}
# sharingMode = maximum, most conservative approach to sharing information across genes to reduce variability of the dispersion estimates
cds <- estimateDispersions(cds, method="pooled", sharingMode="maximum", fitType="parametric")
plotDispEsts(cds)
```

---

## QUALITY CONTROL

### Primary Clustering
- the goal of this analysis is to naiively evaluate the variability within the raw data and determine whether this variability can predict the different sample categories
- Principal Component Analysis  (PCA) is a dimension reduction and visualisation technique that is used to project the multivariate data vector of each sample into a two-dimensional plot, such that the spatial arrangement of the points in the plot reflects the overall data (dis)similarity between the samples.
- in basic terms, we are checking whether the samples cluster by sample type; its a good sign if they do

The plots show a scatterplot of the samples along all dual combinations of the first four principal components. Each sample group is represented by a separate color. You can use these plots to explore if the samples cluster, and whether this is according to an intended experimental factor or according to unintended causes such as batch effects. 

```{r pca, out.width='50%', fig.show='hold'}
PCAplot.cds(cds, normalize.counts=FALSE, title="Raw counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
PCAplot.cds(cds, normalize.counts=TRUE, title="Normalized counts", colorpalette=cbPalette, alpha=0.8, numcomponents=4)
```

---

## DIFFERENTIAL EXPRESSION ANALYSES

### Filtering

Removing genes with no variance in the samples. 

Using this data, I performed independent filtering to eliminate genes that have no, or little chance of showing signicant evidence. Filtering should result in increased detection power, in terms of false discovery rate. Here, genes were filtered  based on the sum of counts from all samples, as below a certain number of counts, it is not possible to get a pvalue below the desired cutoff. I also filtered out any genes without and Ensembl ID annotation.

```{r filter}
## get sum of counts for all samples for each gene
rowcounts <- rowSums(counts(cds, normalized=TRUE))
## filter the data based on the minimal row sum 
use <- (rowcounts > ecdf(rowcounts)(rowcount.cutoff))
cds.filt <- cds[use,]
## annotated?
use <- grepl("ENS", row.names(counts(cds.filt)))
cds.filt <- cds.filt[use,]
```

Here I discarded the genes with less than `r rowcount.cutoff` counts total in all samples, i.e. the bottom `r ecdf(rowcounts)(rowcount.cutoff)*100`% of genes by total counts

### Pairwise Comparisons

Run two pairwise analyses, comparing each FAP sample class to the MEF samples.

```{r pairwise_comparisons, cache=TRUE}
## first construct the actual combinations
all.pair.combos <- combn(as.vector(unique(pData(cds)$condition)),2)
vs.MEF.combos <- all.pair.combos[,which(all.pair.combos[2,]=="MEF")]

# then run the comparisons
nbinom.tests <- alply(all.pair.combos, 2, function(combo) {
  setname <- paste(combo[1], combo[2], sep="-vs-")
  print(setname)
  ## perform significance testing
  allresults <- nbinomTest(cds.filt, combo[1], combo[2])
  ## get normalized counts for significant hits, relabel samples with condition rather than sampleID
  DEresults.counts <- counts(cds.filt, normalize=TRUE)[which(allresults$padj<pvalcutoff & abs(allresults$log2FoldChange)>logfoldcutoff),]
  ## get means and pvalues for significant hits and put together with counts
  DEresults.stats <- allresults[which(allresults$padj<pvalcutoff & abs(allresults$log2FoldChange)>logfoldcutoff),]
  DEresults <- cbind(DEresults.counts, DEresults.stats)
  DEresults <- annotate_df(DEresults)
  DEresults <- DEresults[order(DEresults$padj),]
  ## output some plots
  qqplots <- qq(allresults[,c("pval", "padj")], pvalcutoff, setname)
  DEplots <- plotDE(allresults, pvalcutoff, setname)
  return(list(cond1=combo[1], cond2=combo[2], setname=setname, allresults=allresults, DEresults=DEresults,  qqplots=qqplots, DEplots=DEplots))
})  

nbinom.tests.unfilt <- alply(all.pair.combos, 2, function(combo) {
  setname <- paste(combo[1], combo[2], sep="-vs-")
  print(setname)
  ## perform significance testing
  allresults <- nbinomTest(cds, combo[1], combo[2])
  return(list(allresults=allresults))
  }) 

save.image(file.path(resultsDir, "RDATA.binomtests"))
```

---

## RESULTS

For each comparison, significant genes which passed a `r pvalcutoff*100`% false discovery rate are reported or highlighted.

There are 3 ways to visualize these significant results:
  - QQplot, with significantly varying transcripts shaded in red
  - MA-plot, with significantly varying transcripts shaded in red
  - table of transcripts with significantly different expression levels 

`r setnames <- apply(all.pair.combos, 2, function(n) paste(n[1], n[2], sep=".vs."))`

`r num=1`

### `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]`

There are `r nrow(nbinom.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

```{r out1, out.width='50%', fig.show='hold', dev='png'}
nbinom.tests[[num]]$qqplots
nbinom.tests[[num]]$DEplots
```

**QQ and M-A plots highlighting genes with significant differential expression** 

**Table of top 10 transcripts showing significant differential expression** 

```{r tables1, results='asis'}
if (nrow(nbinom.tests[[num]]$DEresults)>(numsig-1)) {
  out1 <- nbinom.tests[[num]]$DEresults[1:numsig,]
  } else {
    out1  <- nbinom.tests[[num]]$DEresults
    }
out1.gvis <- gvisTable(as.data.frame(apply(out1, 2, as.character)), options=list(width=1280))
print(out1.gvis, "chart")

write.table(nbinom.tests[[num]]$DEresults, file=file.path(resultsDir, paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")), quote=F, sep="\t", row.names=F, col.names=T)
```

[Significant results](../results/deseq/`r paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")`)

`r num=num+1`

### `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]`

There are `r nrow(nbinom.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

```{r out2, out.width='50%', fig.show='hold', dev='png'}
nbinom.tests[[num]]$qqplots
nbinom.tests[[num]]$DEplots
```

**QQ and M-A plots highlighting genes with significant differential expression**  

**Table of top 10 transcripts showing significant differential expression** 

```{r tables2, results='asis'}
if (nrow(nbinom.tests[[num]]$DEresults)>(numsig-1)) {
  out2 <- nbinom.tests[[num]]$DEresults[1:numsig,]
  } else {
    out2  <- nbinom.tests[[num]]$DEresults
    }
out2.gvis <- gvisTable(as.data.frame(apply(out2, 2, as.character)), options=list(width=1280))
print(out2.gvis, "chart")

write.table(nbinom.tests[[num]]$DEresults, file=file.path(resultsDir, paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")), quote=F, sep="\t", row.names=F, col.names=T)
```

[Significant results](../results/deseq/`r paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")`)

`r num=num+1`

### `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]`

There are `r nrow(nbinom.tests[[num]]$DEresults)` genes with significant differential expression (FDR<`r pvalcutoff` and a log2 fold change in expression of at least `r logfoldcutoff`)

```{r out3, out.width='50%', fig.show='hold', dev='png'}
nbinom.tests[[num]]$qqplots
nbinom.tests[[num]]$DEplots
```

**QQ and M-A plots highlighting genes with significant differential expression**  

**Table of top 10 transcripts showing significant differential expression** 

```{r tables3, results='asis'}
if (nrow(nbinom.tests[[num]]$DEresults)>(numsig-1)) {
  out3 <- nbinom.tests[[num]]$DEresults[1:numsig,]
  } else {
    out3  <- nbinom.tests[[num]]$DEresults
    }
out3.gvis <- gvisTable(as.data.frame(apply(out3, 2, as.character)),options=list(width=1280))
print(out3.gvis, "chart")

write.table(nbinom.tests[[num]]$DEresults, file=file.path(resultsDir, paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")), quote=F, sep="\t", row.names=F, col.names=T)
```

[Significant results](../results/deseq/`r paste("DE.genes.q", pvalcutoff, setnames[num], "xls", sep=".")`)

### Count data and statistics for all genes, samples and comparisons

Includes genes which were not significantly DE

```{r output_all, eval=F}
counts.output <- as.data.frame(counts(cds.filt, normalize=TRUE))
counts.output$id <- row.names(counts.output)
counts.output <- annotate_df(counts.output)
setnames <- apply(all.pair.combos, 2, function(x) paste(x, collapse="_vs_"))

for (n in 1:3){
  stats.output <- nbinom.tests[[n]]$allresults
  output <- merge(counts.output,stats.output)
  output <- annotate_df(output)
  output.filename <- file.path(resultsDir, paste("all.exprs.stats", setnames[n], "xls", sep="."))
  write.table(output, file=output.filename, quote=F, sep="\t", row.names=F, col.names=T)
  }

## now forÂ all the genes, not just the filtered genes 
counts.output <- as.data.frame(counts(cds, normalize=TRUE))
counts.output$id <- row.names(counts.output)
for (n in 1:3) {
  stats.output <- nbinom.tests.unfilt[[n]]$allresults
  output <- merge(counts.output,stats.output)
  output <- annotate_df(output)
  output.filename <- file.path(resultsDir, paste("all.exprs.stats.unfiltered.genes", setnames[n], "xls", sep="."))
  write.table(output, file=output.filename, quote=F, sep="\t", row.names=F, col.names=T)  
  }
```

`r num=1`
[Counts and stats for filtered genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](../results/`r  paste("all.exprs.stats.genes", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for filtered genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](../results/`r paste("all.exprs.stats.genes", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for filtered genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](../results/`r paste("all.exprs.stats.genes", setnames[num], "xls", sep=".")`)

`r num=1`
[Counts and stats for ALL  genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](..results/`r paste(".all.exprs.stats.unfiltered.genes", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for ALL genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](..results/`r paste("all.exprs.stats.unfiltered.genes", setnames[num], "xls", sep=".")`)
`r num=num+1`
[Counts and stats for ALL genes and samples in `r all.pair.combos[1,num]` vs. `r all.pair.combos[2,num]` comparison](../results/`r paste("all.exprs.stats.unfiltered.genes", setnames[num], "xls", sep=".")`)

---

## Combining Both Sets of FAP Differentially Expressed Genes

To find genes that are differentially expressed in both FAP cell types as compared to MEFS, we can do a simple search to find genes that are differentially expressed in the same direction (i.e. down in MEFs, up in both FAP cell types) in both of these comparisons.

```{r comboFAP, dev='png', fig.show='hold'}
sig.results.FAPLung.vs.MEF <- nbinom.tests[[2]]$DEresults
sig.results.FAPLung.vs.MEF <- sig.results.FAPLung.vs.MEF[grepl("ENS", sig.results.FAPLung.vs.MEF$id),]

sig.results.FAPMuscle.vs.MEF <- nbinom.tests[[3]]$DEresults
sig.results.FAPMuscle.vs.MEF <- sig.results.FAPMuscle.vs.MEF[grepl("ENS", sig.results.FAPMuscle.vs.MEF$id),]


FAPLung_v_MEF=length(sig.results.FAPLung.vs.MEF$id)
FAPMuscle_v_MEF=length(sig.results.FAPMuscle.vs.MEF$id)
both.sig <- merge(sig.results.FAPLung.vs.MEF[,c("id","log2FoldChange" )], sig.results.FAPMuscle.vs.MEF[,c("id","log2FoldChange" )], by="id", all=FALSE)
both.sig <- both.sig[grepl("ENS", both.sig$id),]
num.both.sig <- length(which(apply(both.sig, 1, function(x) prod(as.numeric(x[2]),as.numeric(x[3]))>0)))

grid.newpage()
draw.pairwise.venn(FAPLung_v_MEF, FAPMuscle_v_MEF, num.both.sig,fill = c(cbPalette[2],cbPalette[3]),
                   alpha = c(0.75, 0.75), lwd=0, cat.dist=0,
                   cat.cex=1, cat.fontfamily = 2,
                   cex = 1, fontfamily =2, 
                   category=c("FAPLung_v_MEF", "FAPMuscle_v_MEF"), 
                   euler.d=T, scaled=T)
grid.newpage()
```

`r idx=2`

For the `r nbinom.tests[[idx]]$setname` comparison, there are `r length(unique(nbinom.tests[[1]]$allresults$id[(nbinom.tests[[idx]]$allresults$padj<pvalcutoff & nbinom.tests[[idx]]$allresults$log2FoldChange<logfoldcutoff)]))` unique genes that are differentially expressed, i.e. with at least `r 2^logfoldcutoff` fold change between the `r nbinom.tests[[idx]]$cond1` and `r nbinom.tests[[idx]]$cond2` sample classes with a pvalue of lower than `r pvalcutoff`.

`r idx=idx+1`

For the `r nbinom.tests[[idx]]$setname` comparison, there are `r length(unique(nbinom.tests[[1]]$allresults$id[(nbinom.tests[[idx]]$allresults$padj<pvalcutoff & nbinom.tests[[idx]]$allresults$log2FoldChange<logfoldcutoff)])
)` unique genes that are differentially expressed, i.e. with at least `r 2^logfoldcutoff` fold change between the `r nbinom.tests[[idx]]$cond1` and `r nbinom.tests[[idx]]$cond2` sample classes.

There are `r num.both.sig` genes that are differentially expressed in the same direction in both FAP cell types as compared to MEFs.

```{r sigoverlap}
#basic hypergeometic overlap
totalassayed <- length(featureNames(cds.filt))
prob.overlap  <- 1 - phyper(num.both.sig - 1, FAPLung_v_MEF, totalassayed - FAPLung_v_MEF, FAPMuscle_v_MEF)
```

Without taking the order of the differentially expressed genes (i.e. how significant or large the changes between samples are), the probability of seeing this many differentially expressed genes in common from both comparisons at random is `r prob.overlap`.

```{r orderlists1}
## Using OrderLists library
# order all  results by log2foldchange
ID.list.FAPLung_v_MEF <- nbinom.tests[[2]]$allresults$id[order(nbinom.tests[[2]]$allresults$log2FoldChange)]
ID.list.FAPMuscle_v_MEF <- nbinom.tests[[3]]$allresults$id[order(nbinom.tests[[3]]$allresults$log2FoldChange)]
## compare with OrderedList, preselect alpha
ol.alpha=0.005
comparedlist <- compareLists(ID.list.FAPLung_v_MEF, ID.list.FAPMuscle_v_MEF, mapping = NULL, two.sided = TRUE, B = 10000, alphas = c(0.1, 0.05, 0.01, 0.005, 0.0025), min.weight = 1e-5, invar.q = 0.5)
getOverlap(comparedlist)
```

`r ol.alpha=0.0025`

We can also compare the lists and take into account that some genes are more differentially expressed than others. The Bioconductor library [OrderedList][5] allows us to examine  overlap of two ordered lists (here ordered by log2 fold change) of differentially expressed genes while giving more weight to the most differentially expressed genes i.e. an overlap between the two gene sets counts more if it is at the top of both lists (most differentially expressed) than the bottom. 
The algorithm progressively moves down the ranks of both lists and calculates the number of genes (overlap) that are present in both lists. These overlaps are summed but each is given less weight as the algorithm progresses through the ranks. The weight given to the observed overlap at each list rank decays exponentially and is determined by the formula exp(-`r ol.alpha` x rank). 

**For example, with the current dataset:**

`r rank=4`
At rank `r rank`, the `r rank` most differentially expressed genes for both comparisons:  
The FAPLung gene list has:   
`r head(ID.list.FAPLung_v_MEF, rank)`  
The FAPMuscle gene list has:  
`r head(ID.list.FAPMuscle_v_MEF, rank)`
`r overlap=length(intersect(head(ID.list.FAPLung_v_MEF, rank), head(ID.list.FAPMuscle_v_MEF,rank)))`
There are `r length(intersect(head(ID.list.FAPLung_v_MEF, rank), head(ID.list.FAPMuscle_v_MEF,rank)))` genes overlapping between the two lists.  
The weight given to any overlap at this rank is: `r exp(-ol.alpha*rank)`  
The weighted overlap is then: `r exp(-ol.alpha*rank)*(as.numeric(overlap))`.  

`r rank=10`
In contrast, at rank `r rank`:
The FAPLung gene list has:   
`r head(ID.list.FAPLung_v_MEF, rank)`  
The FAPMuscle gene list has:  
`r head(ID.list.FAPMuscle_v_MEF, rank)`
`r overlap=length(intersect(head(ID.list.FAPLung_v_MEF, rank), head(ID.list.FAPMuscle_v_MEF,rank)))`
There are `r length(intersect(head(ID.list.FAPLung_v_MEF, rank), head(ID.list.FAPMuscle_v_MEF,rank)))` genes overlapping between the two lists.  
The weight given to any overlap at this rank is: `r exp(-ol.alpha*rank)`  
The weighted overlap is then: `r exp(-ol.alpha*rank)*(as.numeric(overlap))` 

These calculations are repeated after randomly permuting the lists to determine  the chance of this level of overlap occurring at random. Here, the chance of this level of overlap occurring at random is `r getOverlap(comparedlist)$pvalue`.

The a) expression levels and b) differential expression statistics for genes that contribute the most to this overlap and are differentially expressed in both comparisons are listed below.

```{r overlapgenes, results='asis'}
overlapgenes <- getOverlap(comparedlist, max.rank=4605)$intersect

stats.DE.overlapgenes <- merge(nbinom.tests[[2]]$DEresults, nbinom.tests[[3]]$DEresults, by=c('id', 'description', 'mgi_symbol',"FAPLung1", "FAPLung2", "FAPLung3", "FAPLung4", "FAPLung5", "FAPMuscle2", "FAPMuscle3", "FAPMuscle4", "FAPMuscle5", "MEF1", "MEF2", "MEF3", "MEF4"), suffixes=c("_FAPLung.vs.MEF", "_FAPMuscle.vs.MEF"))
stats.overlapgenes <- stats.DE.overlapgenes[stats.DE.overlapgenes$id %in% overlapgenes,]

output.filename <- file.path(resultsDir, "all.exprs.stats.overlapped.comparisons.genes.xls")
write.table(stats.overlapgenes, file=output.filename, quote=F, sep="\t", row.names=F, col.names=T)
```

The following should be the genes that:
1) are differentially expressed (log2FoldChange>`r logfoldcutoff` and adjusted pvalue<`r pvalcutoff`) in both FAP-MEF comparisons and 
2) contribute significantly to the overlap between the two lists of differentially expressed genes from the FAP-MEF comparisons

[Most Commonly Differentially Expressed Genes in Both FAP vs. MEF Comparisons](../results/deseq/all.exprs.stats.overlapped.comparisons.genes.xls)

---

## R Session Information

```{r save_image}
sessionInfo()
save.image(file.path(resultsDir, "RDATA" ))
```


[1]: http://tophat.cbcb.umd.edu (Tophat2)
[2]: http://tophat.cbcb.umd.edu/igenomes.html (iGenomes)
[3]: http://www-huber.embl.de/users/anders/HTSeq/doc/count.html (htseq-count)
[4]: http://www.bioconductor.org/packages/2.12/bioc/html/edgeR.html (edgeR)
[5]: http://www.bioconductor.org/packages/2.12/bioc/html/OrderedList.html (OrderedList)