---
title: "Sahay Lab Klf9 RNA-Seq quality control"
date: "`r BiocStyle::doc_date()`"
author: "Michael J. Steinbaugh"
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE,
                      echo = FALSE,
                      message = FALSE,
                      warning = FALSE)
library(basejump)
library(DESeq2)
library(edgeR)
library(pheatmap)
library(RColorBrewer)
library(readxl)
library(reshape2)
library(tidyverse)
library(tximport)
```

[Klf9]: https://www.omim.org/entry/602902
[R]: https://www.r-project.org
[RStudio]: https://www.rstudio.com

[`bcbio-nextgen`]: https://github.com/chapmanb/bcbio-nextgen
[`bcbio-rnaseq`]: https://github.com/roryk/bcbio.rnaseq
[`DESeq2`]: https://bioconductor.org/packages/release/bioc/html/DESeq2.html
[`ggplot2`]: http://ggplot2.org
[`pheatmap`]: https://cran.r-project.org/web/packages/pheatmap/index.html
[`sailfish`]: http://www.cs.cmu.edu/~ckingsf/software/sailfish
[`tximport`]: https://bioconductor.org/packages/release/bioc/html/tximport.html

```{r bcbio}
# HPC mount point over `sshfs` connection
hpc_dir <- "hpc"
if (!length(dir(hpc_dir))) {
    stop("Project on HPC server not loaded.")
}

# `bcbio-nextgen`
bcbio_dir <- file.path(hpc_dir, "bcbio", "final")

# Default naming scheme is `bcbio/final/YYYY-MM-DD_bcbio`
summary_dir <- file.path(bcbio_dir) %>%
    dir(full.names = TRUE) %>%
    .[grepl("/\\d{4}-\\d{2}-\\d{2}_bcbio$", .)]

# Create directories if necessary
if (!file.exists("data")) {
    dir.create("data")
}
if (!file.exists("results")) {
    dir.create("results")
}
```



# Overview

RNA sequencing experiment performed on dentate gyrus samples from control and transgenic Krueppel-like factor 9 ([Klf9][]) overexpressor mice.

Notes from Kathleen McAvoy:

> All the mice have a dox-inducible drive mDG (rtTA) that activates teto-linked transgenes upon administration of modified dox (9TBD). All mice received 9TBD. The experimental animals have a teto-Klf9 knock-in, which results in increased klf9 mRNA expression in mature DG granule cells (DGCs). A subset of animals in both groups also have a teto-linked Histone 2B-GFP gene, and will display GFP expression. All animals were housed with at least one other cagemate.
> 
> We dissected out the entire dentate gyrus, which includes mainly (likely 80%) mature DG granule cells (DGCs), some other neurons, stem cells (SCs), progenitors, and immature DGCs. The immature DGCs make up a few % of the mDGC population, and will have a different expression profile. The stem and progenitor pool will be even smaller (<1% total) but will again have a very different expression profile. We know that mouse age and our Klf9 overexpression affect the size of the iDGC pool (Klf9 OE doubles it), and both age and Klf9 OE affect stem cell division status and maybe progenitor status (Klf9 OE at least doubles activation). Thus, I tried to match the two groups by age. In most cases, the same age mice were housed together as they were usually littermates. I wouldn't be suprised to see some separation by age, even though the differences are +/- a few weeks. 
> 
> As we know that there are differences in the proportion of some cells and their status, we know that we will likely see global differences in gene expression that are related to the population level, but not necessarily to the single cell level. We'll have to tease this apart in subsequent experiments.

```{r metadata}
metadata <- file.path("data-raw/samples.xlsx") %>%
    readxl::read_excel(.) %>%
    basejump::setNamesSnake(.) %>%
    dplyr::select(name,
                  condition,
                  genotype,
                  cage,
                  age_weeks) %>%
    dplyr::mutate(condition = factor(condition),
                  genotype = factor(genotype),
                  cage = factor(cage),
                  age_weeks = factor(age_weeks))
save(metadata, file = "data/metadata.rda")
write.csv(metadata, file = "results/metadata.csv")
basejump::printTable(metadata, caption = "Samples")
```

```{r summary_data}
summary_data <- file.path(summary_dir, "project-summary.csv") %>%
    readr::read_csv(.) %>%
    basejump::setNamesSnake(.) %>%
    dplyr::select(noquote(order(names(.)))) %>%
    dplyr::mutate(condition = factor(condition)) %>%
    dplyr::arrange(name)
save(summary_data, file = "data/summary_data.rda")
write.csv(summary_data, file = "results/summary_data.csv")
```

```{r txi, include=FALSE}
# `bcbio-nextgen` `sailfish`
tx2gene <- file.path(summary_dir, "tx2gene.csv") %>%
    read.csv(header = FALSE)
sample_dirs <- dir(bcbio_dir) %>%
    .[. %in% summary_data$name] %>%
    file.path(bcbio_dir, .)
sailfish_files <- file.path(sample_dirs, "sailfish", "quant", "quant.sf")
names(sailfish_files) <- metadata$name
txi <- tximport::tximport(sailfish_files,
                          type = "sailfish",
                          tx2gene = tx2gene,
                          reader = readr::read_tsv,
                          countsFromAbundance = "lengthScaledTPM")
save(txi, file = "data/txi.rda")
```

```{r tpm}
# Transcripts per million
tpm <- txi$abundance
save(tpm, file = "data/tpm.rda")
write.csv(tpm, file = "results/tpm.csv")
```

```{r dds, include=FALSE}
# `DESeq2` dataset from `sailfish` output
# Re-adjust for batch effects, if necessary
# ~ genotype + condition + genotype:condition
# ~ condition + age_weeks
dds <- txi %>%
    DESeq2::DESeqDataSetFromTximport(colData = metadata,
                                     design = ~ condition) %>%
    # `betaPrior = FALSE` will default in 2017
    DESeq2::DESeq(.)
DESeq2::resultsNames(dds)
DESeq2::sizeFactors(dds)
# `blind = TRUE` (default) treats all replicates separately
dds_rlog <- DESeq2::rlog(dds)
dds_vst <- DESeq2::varianceStabilizingTransformation(dds)
save(dds, dds_rlog, dds_vst, file = "data/dds.rda")
```

```{r counts}
raw_counts <- DESeq2::counts(dds)
raw_counts_nonzero <- raw_counts %>%
    .[rowSums(.) > 0, ]
normalized_counts <- DESeq2::counts(dds, normalized = TRUE)
normalized_counts_nonzero <- normalized_counts %>%
    .[rowSums(.) > 0, ]
save(raw_counts,
     raw_counts_nonzero,
     normalized_counts,
     normalized_counts_nonzero,
     file = "data/counts.rda")
write.csv(raw_counts, file = "results/raw_counts.csv")
write.csv(normalized_counts, file = "results/normalized_counts.csv")

# `bcbio-nextgen` `featureCounts` (unused)
# raw_counts_featureCounts <- file.path(summary_dir, "combined.counts") %>%
#     read.table(header = TRUE, row.names = 1)
```



# Quality control metrics

## Klf9 gene expression

There's a clear difference in Klf9 expresison between the `control` and `expt` groups. We have re-run the `kmc05` sample as `expt` as requested.

```{r transgene}
plot_gene <- function(gene,
                      counts = tpm,
                      condition = metadata$condition) {
    ylab <- deparse(substitute(counts))
    counts <- as.matrix(counts) %>% .[gene, ]
    df <- data.frame(condition = factor(condition),
                     counts = counts,
                     name = names(counts))
    plot <- ggplot2::ggplot(df,
                            ggplot2::aes(x = name,
                                         y = counts,
                                         fill = condition)) +
        ggplot2::ggtitle(gene) +
        ggplot2::geom_dotplot(binaxis = "y") +
        ggplot2::theme(axis.text.x = element_text(angle = 90)) +
        ggplot2::xlab("sample") +
        ggplot2::ylab(ylab)
    return(plot)
}
plot_gene("ENSMUSG00000033863")
```


## Gender-specific markers

We can evaluate the expression counts of gender-specific markers [@karyagyna2010gender; @staedtler2013gender] to blindly determine gender. Based on these expression profiles, it appears that all the mice studied were male.

```{r gender_markers}
gender_markers <- readr::read_csv("data-raw/gender_markers.csv",
                                  na = c("", "#N/A")) %>%
    dplyr::arrange(chromosome, gene_symbol)
gender_gene_identifier <- gender_markers %>%
    dplyr::filter(include == TRUE) %>%
    dplyr::select(ensembl_gene) %>%
    .[[1]] %>%
    basejump::sortUnique(.)
tpm[gender_gene_identifier, ] %>%
    data.frame %>%
    tibble::rownames_to_column(.) %>%
    reshape2::melt(.) %>%
    setNames(c("ensembl_gene", "name", "counts")) %>%
    dplyr::left_join(gender_markers, by = "ensembl_gene") %>%
    ggplot(aes(x = gene_symbol,
               y = counts,
               color = name,
               shape = chromosome)) +
    ggtitle("Gender-specific markers") +
    geom_jitter(size = 4) +
    xlab("gene") +
    ylab("tpm")
```


## Total reads

Overall read depths for the experiment were good, with all samples having at least 20 million counts.

```{r total_reads}
data.frame(name = summary_data$name,
           total_reads = summary_data$total_reads / 1e6,
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = total_reads,
               fill = condition)) +
    ggtitle("Total reads") +
    geom_bar(stat = "identity") +
    geom_hline(size = 2, yintercept = 20) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("total reads (million)")
```


## Mapped reads

The number of mapped reads corresponds to the number of total reads, as desired.

```{r mapped_reads}
data.frame(name = summary_data$name,
           mapped_reads = summary_data$mapped_reads / 1e6,
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = mapped_reads,
               fill = condition)) +
    ggtitle("Mapped reads") +
    geom_bar(stat = "identity") +
    geom_hline(size = 2,
               yintercept = 20) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("mapped reads (million)")
```


## Genomic mapping rate

The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts. We don't see that here. The mapping rate is acceptable (>90%) for all samples.

```{r genome_mapping_rate}
data.frame(name = summary_data$name,
           genomic_mapping_rate = summary_data$mapped_reads / summary_data$total_reads * 100,
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = genomic_mapping_rate,
               fill = condition)) +
    ggtitle("Genomic mapping rate") +
    geom_bar(stat = "identity") +
    geom_hline(size = 2,
               yintercept = 90) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("genomic mapping rate (%)")
```


## Number of genes detected

The number of genes detected for each sample is good with over 20,000 genes detected for all samples. This is what we expect for the mouse genome.

```{r genes_detected}
data.frame(name = colnames(raw_counts),
           genes_detected = colSums(raw_counts > 0),
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = genes_detected,
               fill = condition)) +
    ggtitle("Number of genes detected") +
    geom_bar(stat = "identity") +
    geom_hline(size = 2,
               yintercept = 20000) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("gene count")
```


## Gene detection saturation

Gene detection saturation was nearly reached for the majority of the samples, which means that most of the samples were able to detect the majority of expressed genes. This is good to see. Also, we observe a linear trend in the number of genes detected with the number of mapped reads, which indicates that the sample input was not overloaded.

```{r saturation}
data.frame(condition = metadata$condition,
           genes_detected = colSums(raw_counts > 0),
           mapped = summary_data$mapped_reads / 1e6) %>%
    ggplot(aes(x = mapped,
               y = genes_detected,
               color = condition,
               fill = condition,
               shape = condition)) +
    ggtitle("Gene detection saturation") +
    geom_point(size = 4) +
    geom_smooth(method = "lm", se = FALSE) +
    xlab("mapped reads (million)") +
    ylab("gene count") 
```


## Exonic mapping rate

The exonic mapping rates are good, with over 60% of total reads mapping to exons.

```{r exonic_mapping_rate}
data.frame(name = summary_data$name,
           exonic_mapping_rate = summary_data$exonic_rate * 100,
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = exonic_mapping_rate,
               fill = condition)) +
    ggtitle("Exonic mapping rate") +
    geom_bar(stat = "identity") +
    geom_hline(size = 2,
               yintercept = 60) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("exonic mapping rate (%)")
```


## rRNA mapping rate

All samples are within an acceptable range (<10%). Samples were purified using the polyA method, which can result in variable rRNA levels from sample to sample. We see higher rRNA mapping rates (>5%) in four samples: `kmc03`, `kmc05`, `kmc10`, `kmc12`.

```{r rrna_mapping_rate}
data.frame(name = summary_data$name,
           rrna_mapping_rate = summary_data$rrna_rate * 100,
           condition = metadata$condition) %>%
    ggplot(aes(x = name,
               y = rrna_mapping_rate,
               fill = condition)) +
    ggtitle("rRNA mapping rate") +
    geom_bar(stat = "identity") +
    geom_hline(linetype = 2,
               yintercept = 5) +
    geom_hline(size = 2,
               yintercept = 10) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab("rRNA mapping rate (%)")
```


## Counts per gene

```{r melt_function}
#' @importFrom reshape2 melt
melt_log10 <- function(counts) {
    counts %>%
        as.data.frame %>%
        reshape2::melt(.,
                       variable.name = "name",
                       value.name = "counts") %>%
        dplyr::filter(counts > 0) %>%
        merge(metadata) %>%
        dplyr::mutate(counts = log(counts)) %>%
        dplyr::arrange(name, -counts)
}
```

Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different. The log10 count spreads look good for all samples.

```{r log10_boxplot}
raw_counts %>%
    melt_log10 %>%
    ggplot(aes(x = name,
               y = counts,
               color = condition)) +
    ggtitle("Counts per gene") +
    geom_boxplot(outlier.shape = NA) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab(expression(log[10]~counts~per~gene))
```

The log10 TMM-normalized counts per gene normalization method [@robinson2010tmm] equates the overall expression levels of genes between samples under the assumption that the majority of them are not differentially expressed. Therefore, by normalizing for total RNA expression by sample, we expect the spread of the log10 TMM-normalized counts per gene to be similar for every sample. The samples look good, with similar log10 TMM-normalized counts per gene distributions.

```{r log10_tmm_boxplot}
tmm_normalized_counts <- raw_counts %>%
    edgeR::DGEList(.) %>%
    edgeR::calcNormFactors(.) %>%
    edgeR::cpm(normalized.lib.sizes = TRUE)
tmm_normalized_counts %>%
    melt_log10 %>%
    ggplot(aes(x = name,
               y = counts,
               color = condition)) +
    ggtitle("Counts per gene (TMM)") +
    geom_boxplot(outlier.shape = NA) +
    theme(axis.text.x = element_text(angle = 90)) +
    xlab("sample") +
    ylab(expression(log[10]~counts~per~gene))
```


## Count density

Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different. The samples look good, with similar log10 TMM-normalized counts per gene densities.

```{r log10_tmm_density}
tmm_normalized_counts %>%
    melt_log10 %>%
    ggplot(aes(x = counts,
               group = name)) +
    ggtitle("Count density (TMM)") +
    geom_density() +
    xlab(expression(log[10]~counts~per~gene)) +
    ylab("density")
```



# Sample similarity analysis

Before performing similarity analysis, we transform counts to log2, which acts to minimize large differences in sequencing depth and helps normalize all samples to a similar dynamic range. For RNA-seq count data, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and instead, low count genes will now dominate the PCA plots. So instead, we use a transformation (`rlog`) that gives similar results for high counts as a log2 transformation but also shrinks the values of low counts towards the genesâ€™ average across samples. We do this with the `rlog` function in the [`DESeq2`][] package [@love2014deseq2], which we will later use for differential gene expression analysis.


## Principal component analysis (PCA)

PCA [@jolliffe2002pca] is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering. In this PCA analysis, we are exploring whether the samples look the same across lanes and whether the genotypes cluster. Unfortunately, the samples do not cluster very well here. This may be indicative of significant uncontrolled variation across samples.

### Grouped by `condition`

```{r pca_condition}
# Consider using `vst` instead of `rlog` for PCA
# `plotPCA()` uses `ntop = 500` by default
# `returnData = FALSE`
dds_rlog %>%
    DESeq2::plotPCA(intgroup = "condition",
                    ntop = nrow(.),
                    returnData = FALSE) +
    ggplot2::ggtitle("PCA, grouped by condition")
```

### Grouped by `genotype`

```{r pca_genotype}
dds_rlog %>%
    DESeq2::plotPCA(intgroup = "genotype",
                    ntop = nrow(.),
                    returnData = FALSE) +
    ggplot2::ggtitle("PCA, grouped by genotype")
```

### Grouped by `age`

The variation in age across samples (8.5-18 weeks) appears to have a significant confounding effect on the study. We see grouping of the samples by age, except the 12-18 age group, which separates into two clusters.

```{r pca_age_weeks}
dds_rlog %>%
    DESeq2::plotPCA(intgroup = "age_weeks",
                    ntop = nrow(.),
                    returnData = FALSE) +
    ggplot2::ggtitle("PCA, grouped by age")
```

### Ungrouped

There appears to be something distinct and perhaps uncontrolled about the `kmc03` sample.

```{r pca_name}
dds_rlog %>%
    DESeq2::plotPCA(intgroup = "name",
                    ntop = nrow(.),
                    returnData = FALSE) +
    ggplot2::ggtitle("PCA, ungrouped")
```


## Hierarchical clustering

Inter-correlation analysis (ICA) is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples. Based on this analysis, it looks like the samples generally have similar transcriptional profiles. Unfortunately, like with the PCA analysis, we don't see tight clustering of the samples by genotype.

```{r heatmap_setup}
heatmap_annotation <- metadata %>%
    dplyr::select(age_weeks,
                  condition,
                  genotype) %>% data.frame
rownames(heatmap_annotation) <- metadata$name
save(heatmap_annotation, file = "data/heatmap_annotation.rda")
heatmap_color <- colorRampPalette(RColorBrewer::brewer.pal(n = 9, name = "Greys"))(100)
```

```{r heatmap_pearson}
dds_rlog %>%
    assay %>%
    cor(method = "pearson") %>%
    pheatmap::pheatmap(annotation = heatmap_annotation,
                       color = heatmap_color,
                       main = "Pearson correlation")
```

```{r heatmap_spearman, include=FALSE}
dds_rlog %>%
    assay %>%
    cor(method = "spearman") %>%
    pheatmap::pheatmap(annotation = heatmap_annotation,
                       color = heatmap_color,
                       main = "Spearman correlation")
```

```{r heatmap_pearson_tmm, include=FALSE}
tmm_normalized_counts %>%
    cor(method = "pearson") %>%
    pheatmap::pheatmap(annotation = heatmap_annotation,
                       color = heatmap_color,
                       main = "Pearson correlation (TMM-normalized)")
```

```{r heatmap_spearman_tmm, include=FALSE}
tmm_normalized_counts %>%
    cor(method = "spearman") %>%
    pheatmap::pheatmap(annotation = heatmap_annotation,
                       color = heatmap_color,
                       main = "Spearman correlation (TMM-normalized)")
```



# Methods

RNA-seq counts were generated by [`bcbio-nextgen`][] and [`bcbio-rnaseq`][] using [`sailfish`][] [@patro2014sailfish]. Counts were imported into [R][]/[RStudio][] using [`tximport`][] [@soneson2015tximport] and [`DESeq2`] [@love2014deseq2]. Plots were generated by [`ggplot2`][] [@wickham2009ggplot2]. Heatmaps were generated by [`pheatmap`][] [@kolde2015pheatmap].


## R session information

`sessionInfo()` output of the workstation that generated this report:

```{r session_info}
sessionInfo()
```



# References
