---
title: "Sahay Lab Klf9 RNA-Seq enrichment"
date: "`r BiocStyle::doc_date()`"
author: "Michael J. Steinbaugh"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE,
                      echo = FALSE,
                      message = FALSE,
                      warning = FALSE)
```

[DAVID]: https://david.ncifcrf.gov
[Gene Ontology]: http://www.geneontology.org
[InterPro]: https://www.ebi.ac.uk/interpro
[KEGG]: http://www.genome.jp/kegg/pathway.html
[UniProt]: http://www.uniprot.org

[`RDAVIDWebService`]: https://bioconductor.org/packages/release/bioc/html/RDAVIDWebService.html

Using only the differentially expressed (DE) genes (`expt_vs_control_de`), we obtained a list of statistically enriched [Gene Ontology][] (GO), [InterPro][], [KEGG][], and [UniProt][] terms from [DAVID][] using the [`RDAVIDWebService`][] interface.

```{r david}
if (!file.exists("data/david.rda")) {
    data(expt_vs_control)
    david <- basejump::enrichDavid(expt_vs_control_de$ensembl_gene_id)
    save(david, file = "data/david.rda")
} else {
    data(david)
}
david$cutoffChart %>%
    dplyr::select(-c(genes, p)) %>%
    basejump::printTable(caption = "DAVID functional annotation chart")
```

```{r session_info}
sessionInfo()
```
