```{r setup, echo=FALSE}
opts_chunk$set(tidy=TRUE, echo=TRUE, highlight=TRUE, figalign="center",
               warning=FALSE, error=FALSE, messages=FALSE, fig.cap="")
```

```{r libraries, warning=FALSE}
library(edgeR)
library(stringr)
library(HTSFilter)
library(googleVis)
library(extrafont)
library(reshape)
#x = font_import(pattern="Gill Sans", prompt=FALSE)
```

# Overview
This dataset has three replicates for each of the following experimental
conditions for 12 total samples.

    1) WT MEF cells
    2) WT MEF cells treated with cholesterol
    3) NRF1 KO cells
    4) NRF1 KO cells treated with cholesterol.

```{r utility-functions}
ensembl_gene = 'mmusculus_gene_ensembl'
gene_symbol = 'mgi_symbol'
filter_type = 'ensembl_gene_id'
annotate_df_for_filtering = function(d) {
	require(biomaRt)
	ensembl = useMart('ensembl', dataset = ensembl_gene)
	a = getBM(attributes=c(filter_type, gene_symbol, "description", "gene_biotype",
                    "chromosome_name"),
		filters=c(filter_type), values=d[, 'id'],
		mart=ensembl)
	m = merge(d, a, by.x="id", by.y=filter_type)
	#names(m)[which(names(m)=="Row.names")]= "id"
	return(m)
}
annotate_df = function(d) {
	require(biomaRt)
	ensembl = useMart('ensembl', dataset = ensembl_gene)
	a = getBM(attributes=c(filter_type, gene_symbol, "description"),
		filters=c(filter_type), values=d[, 'id'],
		mart=ensembl)
	m = merge(d, a, by.x="id", by.y=filter_type)
	#names(m)[which(names(m)=="Row.names")]= "id"
	return(m)
}
mds = function(counts, genotype, treatment) {
        nprobes <- nrow(counts)
        nsamples <- ncol(counts)
        counts = cpm(counts,log=TRUE)
	distances = dist(t(counts))
	fit = cmdscale(distances, eig=TRUE, k=6)
	colnames(fit$points) = c("one", "two", "three", "four", "five", "six")
	df = as.data.frame(fit$points)
	df$label = rownames(df)
        df$genotype = genotype
        df$treatment = treatment
	print(ggplot(df, aes(one, two, color=genotype, shape=treatment))
            + geom_point() +
              theme(text=element_text(family="Gill Sans")))

	eigs = data.frame(variance_explained=fit$eig / sum(fit$eig))
	eigs$component = as.integer(rownames(eigs))
	#print(ggplot(eigs, aes(component, variance_explained)) + geom_point())
}

bcv = function(y) {
#       Check y
        if(!is(y,"DGEList")) stop("y must be a DGEList.")

#       Compute AveLogCPM if not found in y
        A <- aveLogCPM(y$counts,offset=getOffset(y))

#       Points to determine y axis limits
        disp <- sqrt(getDispersion(y))
        if(is.null(disp)) stop("No dispersions to plot")
        if(attr(disp,"type")=="common") disp <- rep(disp,length=length(A))
        df = data.frame(A, disp)

#       Make plot
        p = ggplot(df, aes(A, disp)) + geom_point(alpha=1/10)
        labels <- cols <- NULL
         if(!is.null(y$tagwise.dispersion)) {
           df$tagwise.dispersion = sqrt(y$tagwise.dispersion)
           p = p + geom_point(data=df, aes(A, tagwise.dispersion), alpha=1/10)
         }
        if(!is.null(y$common.dispersion)) {
           df$common.dispersion = sqrt(y$common.dispersion)
           p = p + geom_line(data=df, aes(A, common.dispersion), color="red")
        }
        if(!is.null(y$trended.dispersion)) {
           o <- order(A)
           df$trended.dispersion = y$trended.dispersion
           x = df[o]
           p = p + geom_line(data=x, aes(A, trended.dispersion), color="blue")
}
        p + xlab("Average log CPM") +
            ylab("Biological coefficient of variation") +
            theme(text=element_text(family="Gill Sans"))
}

annotate_go = function(de_genes, assayed_genes) {
  library(goseq)
  library(GO.db)
  genes = assayed_genes %in% de_genes
  names(genes) = assayed_genes
  pwf = nullp(genes, "mm9", "ensGene", plot.fit=FALSE)
  GO.wall = goseq(pwf, "mm9", "ensGene")
  GO.wall$fdr = p.adjust(GO.wall$over_represented_pvalue, method="BH")
  enriched = subset(GO.wall, fdr < 0.10)
  GO_names = c()
  for (go in enriched$category) {
    GO_names = c(GO_names, GOTERM[[go]]@Term)
  }
  enriched$GO_name = GO_names
  enriched$pvalue = enriched$over_represented_pvalue
  enriched = enriched[, c("category", "GO_name", "pvalue", "fdr")]
  return(enriched)
}
```

First load the data and do some basic cleaning and setup. We are not going to be
able to quantitate non-polyA selected rRNAs very well, so we will remove
them from the analysis since they are noisy. We also set up a the basic
groups for the experiment for pairwise comparisons as well as a design
matrix for a more complicated GLM-based treatment.
```{r data-load}
count_file = "/Users/rory/hsph/hsph/projects/widenmaier_rnaseq/bcbio_test/htseq-count/combined.counts"
counts = read.table(count_file, header=TRUE, row.names="id")
# remove the non-aligned counts
counts = head(counts, -5)
# remove counts for anything on MT or not protein coding/pseudogene
tmp = counts
tmp$id = rownames(counts)
tmp = annotate_df_for_filtering(tmp)
tmp = subset(tmp, gene_biotype %in% c("psuedogene", "protein_coding"))
tmp = subset(tmp, chromosome_name != "MT")
counts = counts[tmp$id,]
sample = colnames(counts)
short_names = unlist(lapply(sample,
                    function(x) str_split(x, ".querysorted")[[1]][1]))
colnames(counts) = short_names
cell_type = unlist(lapply(sample, function(x) str_split(x, "_")[[1]][1]))
condition = unlist(lapply(sample, function(x) str_split(x, "_")[[1]][2]))
replicate = unlist(lapply(sample, function(x) str_split(x, "_")[[1]][3]))
replicate = unlist(lapply(replicate,
                    function(x) str_split(x, ".querysorted")[[1]][1]))
group = data.frame(sample=short_names, cell_type=cell_type, condition=condition,
  replicate=replicate)
group$cell_type = relevel(group$cell_type, ref="WT")
group$condition = relevel(group$condition, ref="Con")
design = model.matrix(~ condition + cell_type + condition:cell_type, data=group)
```
<!---
#edgeR.dgelist = DGEList(counts = counts)
#edgeR.dgelist = calcNormFactors(edgeR.dgelist, method = "TMM")
#edgeR.dgelist = estimateGLMCommonDisp(edgeR.dgelist, design)
#edgeR.dgelist = estimateGLMTrendedDisp(edgeR.dgelist, design)
#edgeR.dgelist = estimateGLMTagwiseDisp(edgeR.dgelist, design)
#fit = glmFit(edgeR.dgelist, design)
#lrt = glmLRT(fit)
#topTags(lrt)

#de_table = lrt$table
#de_table$DE = decideTests(de_table)
#de_table$FDR = p.adjust(de_table$PValue, method="BH")
#de_table$id = rownames(de_table)
#x = annotate_df(de_table)
--->

```{r pairwise-setup}
pairwise = paste(group$cell_type, group$condition, sep="_")
edgeR.dgelist = DGEList(counts = counts, group=pairwise)
```

# Normalization
We use the edgeR TMM (trimmed mean of m-values) to perform normalization of the
count data to allow comparisons across conditions and within conditions.

```{r normalization}
edgeR.dgelist = calcNormFactors(edgeR.dgelist, method = "TMM")
```

```{r pre-normalization}
tmp_cpm = data.frame(cpm(edgeR.dgelist, normalized.lib.sizes=FALSE))
q = tmp_cpm/rowMeans(tmp_cpm)
binwidth = diff(range(q, na.rm=TRUE))/100
q$genes = rownames(q)
q = melt(q, id='genes', variable_name='sample')
q = merge(q, group, by.x="sample", by.y="sample")
ggplot(q, aes(value), environment=environment()) +
  geom_histogram(binwidth=0.01) +
  scale_x_log10() + facet_grid(cell_type + condition ~ .) +
  ggtitle("log10 counts, pre-TMM") +
  xlab("log10 counts") +
  theme(text=element_text(family="Gill Sans"), axis.title.y=element_blank())
```
```{r post-normalization}
tmp_cpm = data.frame(cpm(edgeR.dgelist))
q = tmp_cpm/rowMeans(tmp_cpm)
binwidth = diff(range(q, na.rm=TRUE))/100
q$genes = rownames(q)
q = melt(q, id='genes', variable_name='sample')
q = merge(q, group, by.x="sample", by.y="sample")
ggplot(q, aes(value), environment=environment()) +
  geom_histogram(binwidth=0.01) +
  scale_x_log10() + facet_grid(cell_type + condition ~ .) +
  ggtitle("log10 counts, post-TMM") +
  xlab("log10 counts") +
  theme(text=element_text(family="Gill Sans"), axis.title.y=element_blank())
```

Normalizing the data causes fairly nice clusters in the MDS plot, with
samples roughly clustering based on cell type and exposure to cholesterol.
```{r normalization-mds, fig.cap="MDS plot"}
mds(counts, cell_type, condition)
```

A heatmap of the correlations between the sample seems to support this
as well, with cells in the same condition mostly having higher
correlation than across conditions.
```{r heatmap-correlation}
qplot(x=X1, y=X2, data=melt(cor(log(edgeR.dgelist$counts + 1))),
    fill=value, geom="tile") + theme(axis.text.x=element_text(angle=-90),
    text=element_text(family="Gill Sans"), axis.title=element_blank()) +
    scale_fill_continuous(name="Pearson correlation")
```

# Dispersion estimation
```{r dispersion}
edgeR.dgelist = estimateCommonDisp(edgeR.dgelist)
#edgeR.dgelist = estimateTrendedDisp(edgeR.dgelist)
edgeR.dgelist = estimateTagwiseDisp(edgeR.dgelist, trend = "movingave")
```
The biological coefficient of variation (BCV) is < 20%, which is pretty typical
of an experiment involving either a cell line or a model organism. For
genes with low counts.
```{r bcv-plot}
bcv(edgeR.dgelist)
```
The red line is the common dispersion; if you were to use the dispersions
of all genes as the variance to test against you can see that it would
underestimate the dispersion of lowly expressed genes and overestimate
the dispersion of many genes more highly expressed. There are strategies
which use a trended dispersion, so doing a moving average of the dispersion,
which works much better than using a single value but still is inaccurate
for some genes. Since we have replicates, we can use a tagwise dispersion
estimation which uses both the common dispersion and the dispersion calculated
from individual genes to calculate the dispersion of a gene, splitting
the difference between the common and tag dispersions. With more replicates,
the dispersion estimation is pulled more towards the tagwise dispersion
for each gene and away from the common dispersion.

# Filtering
To maximize power of the statistical tests, we want to eliminate
genes from consideration that have either no signal or a low,
constant, uninformative signal across experimental conditions. We choose
a threshold maximizing, between each set of replicates, sum of Jaccard index
for all pairwise comparisons of replicates.

```{r filtering}
et = exactTest(edgeR.dgelist, pair=c("WT_Con", "KO_Con"))
et = HTSFilter(et, DGEList=edgeR.dgelist, s.len=25)$filteredData
bcv(edgeR.dgelist[rownames(et$table),])
```
Filtering removes a lot of the high variance genes which are not strongly
expressed across samples and which, on a whole, provide very little
useful differential expression information.

# WT control vs KO control
```{r WT_vs_KO, results='asis'}
et = exactTest(edgeR.dgelist, pair=c("WT_Con", "KO_Con"))
et = HTSFilter(et, DGEList=edgeR.dgelist, s.len=25, plot=FALSE)$filteredData
et_table = et$table
et_table$FDR = p.adjust(et_table$PValue, method="BH")
de_table = subset(et_table, FDR < 0.10)
de_table$id = rownames(de_table)
WT_vs_KO = annotate_df(de_table)
print(gvisTable(WT_vs_KO), "chart", options=list(width=900))
write.table(WT_vs_KO, "WT_vs_KO_de.txt", quote=FALSE, row.names=FALSE, sep="\t")
```

# WT cholesterol treated vs KO cholesterol treated
```{r WT_chol_vs_KO_chol, results='asis'}
et = exactTest(edgeR.dgelist, pair=c("WT_Chol", "KO_Chol"))
et = HTSFilter(et, DGEList=edgeR.dgelist, s.len=25, plot=FALSE)$filteredData
et_table = et$table
et_table$FDR = p.adjust(et_table$PValue, method="BH")
de_table = subset(et_table, FDR < 0.10)
de_table$id = rownames(de_table)
WT_chol_vs_KO_chol = annotate_df(de_table)
print(gvisTable(WT_chol_vs_KO_chol), "chart", options=list(width=900))
write.table(WT_chol_vs_KO_chol, "WT_chol_vs_KO_chol_de.txt", quote=FALSE,
  row.names=FALSE, sep="\t")
```

# WT control vs WT cholesterol treated
```{r WT_Con_vs_WT_chol, results='asis'}
et = exactTest(edgeR.dgelist, pair=c("WT_Con", "WT_Chol"))
et = HTSFilter(et, DGEList=edgeR.dgelist, s.len=25, plot=FALSE)$filteredData
et_table = et$table
et_table$FDR = p.adjust(et_table$PValue, method="BH")
de_table = subset(et_table, FDR < 0.10)
de_table$id = rownames(de_table)
WT_vs_WT_chol = annotate_df(de_table)
print(gvisTable(WT_vs_WT_chol), "chart", options=list(width=900))
write.table(WT_vs_WT_chol, "WT_vs_WT_chol_de.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```

# KO control vs KO cholesterol treated
```{r KO_Con_vs_KO_chol, results='asis'}
et = exactTest(edgeR.dgelist, pair=c("KO_Con", "KO_Chol"))
et = HTSFilter(et, DGEList=edgeR.dgelist, s.len=25, plot=FALSE)$filteredData
et_table = et$table
et_table$FDR = p.adjust(et_table$PValue, method="BH")
de_table = subset(et_table, FDR < 0.10)
de_table$id = rownames(de_table)
KO_vs_KO_chol = annotate_df(de_table)
print(gvisTable(KO_vs_KO_chol), "chart", options=list(width=900))
write.table(KO_vs_KO_chol, "KO_vs_KO_de.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```

# GO ontology analysis
```{r go-ontology-example}
library(goseq)
library(GO.db)
assayed.genes = rownames(edgeR.dgelist$counts)
genes = rownames(edgeR.dgelist$counts) %in% KO_vs_KO_chol$id
names(genes) = assayed.genes
```
Larger genes are more likely to be called DE, so we have to correct for that
in the ontology analysis:
```{r gene-length-de-bias}
    pwf = nullp(genes, "mm9", "ensGene")
```

## WT control vs KO control treated GO analysis
```{r go-WT_vs_KO, results='asis'}
enriched = annotate_go(WT_vs_KO$id, rownames(edgeR.dgelist$counts))
print(gvisTable(enriched), "chart", options=list(width=900))
write.table(enriched, "WT_vs_KO_go.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```
## WT cholesterol vs KO cholesterol treated GO analysis
```{r go-WT_chol_vs_KO_chol, results='asis'}
enriched = annotate_go(WT_chol_vs_KO_chol$id, rownames(edgeR.dgelist$counts))
print(gvisTable(enriched), "chart", options=list(width=900))
write.table(enriched, "WT_chol_vs_KO_chol_go.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```
## WT control vs WT cholesterol treated GO analysis
```{r go-WT_vs_WT_chol, results='asis'}
enriched = annotate_go(WT_vs_WT_chol$id, rownames(edgeR.dgelist$counts))
print(gvisTable(enriched), "chart", options=list(width=900))
write.table(enriched, "WT_vs_WT_chol_go.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```
## KO control vs KO cholesterol treated GO analysis
```{r go-KO_vs_KO_chol, results='asis'}
enriched = annotate_go(KO_vs_KO_chol$id, rownames(edgeR.dgelist$counts))
print(gvisTable(enriched), "chart", options=list(width=900))
write.table(enriched, "KO_vs_KO_chol_go.txt", quote=FALSE, row.names=FALSE,
  sep="\t")
```
