---
title: "450K CpG probes near miRNAs"
output:
  html_document:
    toc: true
    theme: spacelab
---
```{r setup, echo=FALSE}
library(knitr)
opts_chunk$set(tidy=TRUE, cache=FALSE,  highlight=TRUE, figalign="center", echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, tidy.opts=list(keep.blank.line=FALSE, width.cutoff=120))
options(width=200)
```



# Load libraries
```{r libraries}
library(biomaRt)
library(knitr)
library(stringr)
library(xlsx)
library(ggplot2)
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
library(rtracklayer)

```

# Functions
```{r functions}
bedTools.2in<-function(functionstring="bedtools intersect",bed1,bed2,opt.string="") {
  #create temp files
  a.file=tempfile()
  b.file=tempfile()
  out   =tempfile()
  options(scipen =99) # not to use scientific notation when writing out
 
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  write.table(bed2,file=b.file,quote=F,sep="\t",col.names=F,row.names=F)
 
  # create the command string and call the command using system()
  command=paste(functionstring,"-a",a.file,"-b",b.file,opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
 
  res=read.table(out,header=F)
  unlink(a.file);unlink(b.file);unlink(out)
  return(res)
}
bedTools.slop <- function(functionstring="bedtools slop", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  # create the command string and call the command using system()
  command <- paste(functionstring,"-i",a.file,"-g", genomefile, opt.string,">",out,sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

bedTools.flank <- function(functionstring="bedtools flank -s", genomefile, bed1, opt.string="") {
  #create temp files
  a.file <- tempfile()
  out  <- tempfile()
  options(scipen=99) # not to use scientific notation when writing out  
  #write bed formatted dataframes to tempfile
  write.table(bed1,file=a.file,quote=F,sep="\t",col.names=F,row.names=F)
  
  # create the command string and call the command using system()
  command <- paste(functionstring,"-g", genomefile, opt.string, "-i", a.file, ">", out, sep=" ")
  cat(command,"\n")
  try(system(command))
  
  res <- read.table(out,header=F)
  unlink(a.file);unlink(out)
  return(res)
  }

```


# Set directories

```{r directories}
baseDir <- "~/projects/ab_nucmito/"
dataDir <- file.path(baseDir, "data")
resultsDir <- file.path(baseDir, "results")
metaDir <- file.path(baseDir, "meta")
```

# Load data
- miRNA data for homo sapiens downloaded from mirBase FTP site (http://www.mirbase.org/ftp.shtml), mirBase build 19 for GRCh37

```{r dataload, cache=TRUE}
annots.450k <- read.csv(file.path(dataDir, "14_April_2014_Updated_humanmethylation450_15017482_v1-2.csv"), skip=7)
miRNA <- import.gff3(file.path(dataDir, "hsa.gff3.txt"))
```

# Subset miRNAs to primary transcripts on known chromosomes  
- subset to primary transcripts
- convert to [BED format](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)
- dicard all mitochondrial, patches and polymorphic regions
- discard any duplicates

```{r subset}
miRNA <- miRNA[which(miRNA$type=="miRNA_primary_transcript"),]

miRNA <- data.frame(chr=seqnames(miRNA),
  start=start(miRNA)-1,
  end=end(miRNA),
  name=miRNA$Name,
  score=c(rep(".", length(miRNA))),
  strand=strand(miRNA))
chrs <- paste("chr", c(seq(1,22), "X", "Y"), sep="")
miRNA <- miRNA[miRNA$chr %in% chrs,]
# dedupe
miRNA <- miRNA[with(miRNA, order(chr,start, end)),]
miRNA <- miRNA[!duplicated(miRNA[,c("start", "end", "chr")]),]
```

# Get coordinates of regions around miRNAs
3kb upstream of the TSS and 3Kb downstream of transcript end

```{r getpromoters}
miRNA.flanked <- bedTools.slop(bed1 = miRNA, opt.string = "-b 3000", genomefile = "~/projects/ab_nucmito/meta/hg19.genome")
miRNA.upstream <- bedTools.flank(bed1 = miRNA, opt.string = "-l 3000 -r 0", genomefile = "~/projects/ab_nucmito/meta//hg19.genome")
miRNA.downstream <- bedTools.flank(bed1 = miRNA, opt.string = "-l 0 -r 3000", genomefile = "~/projects/ab_nucmito/meta//hg19.genome")
```

# Find 450K probes within flanks and genes of miRNAs

## Extract coordinates (Human Genome build GRCh37) of 450K probes from annotation file and convert to BED format
- remove non-CpG probes

```{r 450kextract}
bed.450K <- annots.450k[,c("CHR", "MAPINFO", "Name")]
# remove non-CpG probes

bed.450K <- bed.450K[grep("cg", bed.450K$Name),]
# adjust to zero-based coordinate format
bed.450K$start <- bed.450K$MAPINFO-1
names(bed.450K) <- c("chr", "end", "name", "start")
bed.450K$chr <- paste("chr", bed.450K$chr, sep="")
bed.450K <- bed.450K[,c("chr", "start", "end", "name")]
```

## Overlap 450K probe coordinates with flanks of miRNAs' gene coordinates

```{r overlap}
probes450K.in.flanked.miRNAs <- bedTools.2in(bed1=bed.450K,bed2=miRNA.flanked,opt.string="-wo")
probes450K.in.flanked.miRNAs <- probes450K.in.flanked.miRNAs[,c(1,2,3,4,8)]
names(probes450K.in.flanked.miRNAs) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.flanked.miRNAs <- probes450K.in.flanked.miRNAs[with(probes450K.in.flanked.miRNAs, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.flanked.miRNAs <- probes450K.in.flanked.miRNAs[!duplicated(probes450K.in.flanked.miRNAs),]

##########################
##########################
probes450K.in.upstream.miRNAs <- bedTools.2in(bed1=bed.450K,bed2=miRNA.upstream,opt.string="-wo")
probes450K.in.upstream.miRNAs <- probes450K.in.upstream.miRNAs[,c(1,2,3,4,8)]
names(probes450K.in.upstream.miRNAs) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.upstream.miRNAs <- probes450K.in.upstream.miRNAs[with(probes450K.in.upstream.miRNAs, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.upstream.miRNAs <- probes450K.in.upstream.miRNAs[!duplicated(probes450K.in.upstream.miRNAs),]

##########################
##########################

probes450K.in.downstream.miRNAs <- bedTools.2in(bed1=bed.450K,bed2=miRNA.downstream,opt.string="-wo")
probes450K.in.downstream.miRNAs <- probes450K.in.downstream.miRNAs[,c(1,2,3,4,8)]
names(probes450K.in.downstream.miRNAs) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.downstream.miRNAs <- probes450K.in.downstream.miRNAs[with(probes450K.in.downstream.miRNAs, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.downstream.miRNAs <- probes450K.in.downstream.miRNAs[!duplicated(probes450K.in.downstream.miRNAs),]

##########################
##########################

probes450K.in.miRNAs <- bedTools.2in(bed1=bed.450K,bed2=miRNA,opt.string="-wo")
probes450K.in.miRNAs <- probes450K.in.miRNAs[,c(1,2,3,4,8)]
names(probes450K.in.miRNAs) <- c("chr", "start", "end", "probeID", "overlapped_gene")
# sort and remove duplicates
probes450K.in.miRNAs <- probes450K.in.miRNAs[with(probes450K.in.miRNAs, order(chr, start, probeID, overlapped_gene)), ]
probes450K.in.miRNAs <- probes450K.in.miRNAs[!duplicated(probes450K.in.miRNAs),]

```

# Output results

```{r output}
setwd(resultsDir)
write.xlsx2(probes450K.in.flanked.miRNAs, file="probes450K.in.flanked.miRNAs.xlsx")
write.xlsx2(probes450K.in.upstream.miRNAs, file="probes450K.upstream.of.miRNAs.xlsx")
write.xlsx2(probes450K.in.downstream.miRNAs, file="probes450K.downstream.of.miRNAs.xlsx")
write.xlsx2(probes450K.in.miRNAs, file="probes450K.in.miRNAs.xlsx")
```

##[Excel file with all results](../results/probes450K.in.flanked.miRNAs.xlsx)
##[Excel file with upstream results](../results/probes450K.upstream.of.miRNAs.xlsx)
##[Excel file with downstream results](../results/probes450K.downstream.of.miRNAs.xlsx)
##[Excel file within miRNA results](../results/probes450K.in.miRNAs.xlsx)

# Plot categorical results

```{r plotcategories}

num.genic <- length(unique(probes450K.in.miRNAs$probeID))
num.upstream <- length(unique(probes450K.in.upstream.miRNAs$probeID))
num.downstream <- length(unique(probes450K.in.downstream.miRNAs$probeID))

nums <- as.data.frame(c(num.upstream,num.genic, num.downstream))
names(nums) <- "count"
nums$category <- c("upstream","genic",  "downstream")
nums <- within(nums, category <- factor(category,levels=c("upstream", "genic","downstream"))) # rearrange factor levels for plot


# plot bars and add text
p <- ggplot(nums, aes(x = category, y = count, fill=factor(category))) +
  geom_bar(stat="identity")+
  scale_fill_manual(values=cbPalette)+
  theme_bw()+ 
  theme(legend.title=element_blank())+
  theme(axis.title.x = element_blank())
p+ geom_text(data=nums,aes(x=category,y=0.5*count,label=count), size=2.5) 

```

